Fix issues related to hr tags

facelessuser opened the pull request:
Ensure that start/end tag handler does not include tags in the previous
paragraph.

Provide special handling for tags like hr that never have content.

facelessuser committed:
Fix issues related to hr tags

Ensure that start/end tag handler does not include tags in the previous
paragraph.

Provide special handling for tags like hr that never have content.

facelessuser commented:
Needs tests and is a work in progress. Mainly something that can be discussed.

waylan commented on markdown/htmlparser.py:
```diff
@@ -56,6 +56,10 @@ class HTMLExtractor(htmlparser.HTMLParser):
             kwargs['convert_charrefs'] = False
+
+        # Block tags that should contain no content (self closing)
+        self.empty_tags = ['hr']
```
You seem to think we need to treat `<hr>` specially. However, I would expect all 'block-level tags' to behave the same way. Therefore `<script />` would be no different. We only need to check if a tag is a block-level tag in `handle_empty_tags`. No need to specifically list empty tags separately.

facelessuser replied:
`<hr>` is a different case than `<script />`. The latter is giving a hint that it is self closed.

`<hr>` gives no hint, and usually, you would keep looking for an ending tag. But `<hr>` *never* has content. Maybe I'm missing something, but this is a very special case.

facelessuser replied:
In fact, given:

```
<hr>
some text
</hr>
```

You should treat them both as separate `<hr>` like browsers do.

waylan replied:
I'm operating on the assumption that the HtmlParser properly recognizes that `<hr>` is a self-closing tag. If that is not the case, then that might be different.

waylan replied:
Sure enough, the HtmlParser does not recognize `<hr>` as a self-closing tag.

```python
>>> from test.test_htmlparser import EventCollector
>>> ec = EventCollector()
>>> ec.feed('<hr>')
>>> ec.close()
>>> ec.get_events()
[('starttag', 'hr', [])]
>>> ec = EventCollector()
>>> ec.feed('<hr/>')
>>> ec.close()
>>> ec.get_events()
[('startendtag', 'hr', [])]
```

waylan replied:
So I see that `<hr />` fails in a different way:

```html
<p>foo
    <hr />
</p>
<p>bar</p>
```

Sigh. Apparently it never occurred to me we would need separate tests specifically for `hr` tags.

waylan commented on markdown/htmlparser.py:
```diff
@@ -264,6 +272,14 @@ def parse_starttag(self, i):  # pragma: no cover
         if end.endswith('/>'):
             # XHTML-style empty tag: <span attr="value" />
             self.handle_startendtag(tag, attrs)
+        elif tag in self.empty_tags:
```
There should be no changes to the `parse_starttag` method. We only override this method to fix the upstream issue. In all other respects this is an exact copy of the parent class. See the comment on this method would suggest the same. We need to address the issue in the public functions first if we can.

waylan commented on markdown/extensions/md_in_html.py:
```diff
@@ -36,9 +36,10 @@ def __init__(self, md, *args, **kwargs):
-        self.block_tags = [tag for tag in self.block_level_tags if tag not in self.span_tags + self.raw_tags]
         super().__init__(md, *args, **kwargs)
 
+        self.block_tags = [tag for tag in self.block_level_tags if tag not in self.span_tags + self.raw_tags + self.empty_tags]
```
Btw, I see you first tried to use sets here. The appropriate way to do that would be:

```python
self.block_tags = set(self.block_level_tags) - set(self.span_tags + self.raw_tags + self.empty_tags)
```

I debated which way to go with it. In the end I chose to keep using lists rather than changing to sets. I did not check performance. If sets are faster, then by all means, feel free to make the change.

facelessuser committed:
Tests and changes

- Add tests
- Move new logic out of parse_starttag
- Use sets for block tag lists as they are much faster when comparing
  if an item is in the list.

facelessuser commented:
I've added tests. For cases with ending `</hr>` tags, I'm not overly concerned with the unexpected behavior. One should never use ending `</hr>` tags. I was examining browser behavior, and they seem to ignore them. `<hr>` and `<hr/>` should get treated in a sane manner though.

facelessuser committed:
Update changelog

facelessuser commented:
This took me longer than I planned. A second set of eyes to double check I'm not missing something may be good. Let me know if there are any test cases I'm missing.

Hopefully, the approach is fine. It took me a bit to massage it into a state where I was able to handle our failing cases, but not break other cases.

waylan commented on tests/test_syntax/blocks/test_html_blocks.py:
```diff
@@ -1402,3 +1402,67 @@ def test_inline_script_tags(self):
+
+    def test_hr_only_start(self):
+        self.assertMarkdownRenders(
+            "*emphasis1*\n<hr>\n*emphasis2*",
```
I find these condensed strings a little more difficult to parse (as a human) than I like. Can we please use `dedent` and string literals? It is much easier to read that way.

waylan commented on tests/test_syntax/blocks/test_html_blocks.py:
```diff
@@ -1402,3 +1402,67 @@ def test_inline_script_tags(self):
+            "*emphasis1*\n<hr></hr>\n*emphasis2*",
+            self.dedent(
+                """
+                <p><em>emphasis1</em></p>
+                <hr>
+                <p></hr>
+                <em>emphasis2</em></p>
```
This one is surprising to me. I would expect:

```html
<p><em>emphasis1</em></p>
<hr>
</hr>
<p><em>emphasis2</em></p>
```

I realize the `/` is at the start of the tag, not the end. I suppose the browser probably is weird about this. I might need to think some more about this.

waylan replied:
> One should never use ending `</hr>` tags. I was examining browser behavior, and they seem to ignore them. 

You're right. They are completely ignored by browsers. That being the case, then I guess it doesn't matter how we handle them. Perhaps we should add a comment on this test about browser behavior.

waylan reviewed (changes requested):
This is looking good. I have a couple notes below.

facelessuser committed:
Add comments about end hr and use dedent in tests
